### ISSUE TITLE
Add type annotations to the codebase

### ISSUE THREAD

**Issue by akihironitta** [PyTorchLightning/pytorch-lightning-bolts#434]
## 🚀 Feature
As static type checker `mypy` was added in #433, we can add type annotations to the codebase.

**Comment by hassiahk** [comment]
Any help needed with this issue? I am happy to help. 😄

**Comment by akihironitta** [comment]
@hassiahk Yes! I just added the steps to contribute to the above comment, so it'd be great if you follow them and submit a PR! :]

**Comment by hassiahk** [comment]
@akihironitta, it would be great if you can also add the below to the initial comment so that we can keep track of which ones are done and this will also ensure no two people are clashing with each other's work. Thanks!!

### Sections
- [ ] `pl_bolts.callbacks.*` <---| @ydcjeff 
- [ ] `pl_bolts.datamodules.*`
- [ ] `pl_bolts.datasets.*`
- [ ] `pl_bolts.losses.*`
- [ ] `pl_bolts.metrics.*`
- [ ] `pl_bolts.models.*`
- [ ] `pl_bolts.models.autoencoders.*`
- [ ] `pl_bolts.models.detection.*`
- [ ] `pl_bolts.models.gans.*`
- [ ] `pl_bolts.models.regression.*`
- [ ] `pl_bolts.models.rl.*`
- [ ] `pl_bolts.models.self_supervised.*`
- [ ] `pl_bolts.models.vision.*`
- [ ] `pl_bolts.optimizers.*`
- [ ] `pl_bolts.transforms.*`
- [ ] `pl_bolts.utils.*` <---| @hassiahk 
- [ ] `tests.*`

**Comment by akihironitta** [comment]
@hassiahk Done :]

**Comment by hassiahk** [comment]
@akihironitta, should I wait for https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/450 to get merged? or can I install `mypy` and proceed with this issue?

**Comment by akihironitta** [comment]
@hassiahk Sure, you can start to work on this issue!

**Comment by briankosw** [comment]
Hey @akihironitta, I'd love to help with this :) Should I just pick on and create a PR?

**Comment by akihironitta** [comment]
@briankosw Yes! Thank you!

**Comment by hassiahk** [comment]
Hey @briankosw. Yes, you can pick one and create a PR for it. 😄 Just leave a comment here saying you picked up this, so you won't clash with others.

**Comment by briankosw** [comment]
I guess I'll work on `pl_bolts.datamodules` then!

**Comment by hassiahk** [comment]
@akihironitta, I will be working on `losses` next. :)

**Comment by johnpeterson123** [comment]
I'm going to work on `pl_bolts.optimizers` now!

**Comment by akihironitta** [comment]
@johnpeterson123 Thank you for your contribution!

**Comment by jeromepatel** [comment]
I am interested in working on `pl_bolts.models.regression.*` .

**Comment by akihironitta** [comment]
Not sure if we want to continue on this issue because there will be a lot of `# type: ignore` comments in the codebase (like in #444 and #608), which I think lowers readability. Should we still keep adding type hints to Bolts?
